tensorflow/python/ops/array_ops.py

```diff
@@ -3532,7 +3549,114 @@ def _get_paddings_constant(paddings):
     return [_get_paddings_constant(x) for x in paddings]
   else:
     return paddings
+  
+def _wrap_pad(tensor, paddings, name=None):
+  """Helper to create wrap pad option with ops already built.
 
+  See `pad` for more info on how this function works.
+  """
+  # Use constant padding to create a tensor with final shape and original values.
+  padded = gen_array_ops.pad(input=tensor, paddings=paddings, name=name)
+  start_coords, end_coords = [
+    shape(padded) - paddings[:,1] - shape(tensor), 
+    shape(padded) - paddings[:,1]
+    ]
+  tensor_to_copy = tensor
+  strides = ones(rank(padded))
+  # Add wraps for each dimension.
+  for i in range(shape(paddings)[0]):
+    # The shape of `tensor_to_copy` can change in iterations. 
+    # Set the loop options to avoid an error in graph 
+    # mode.
+    directives.set_loop_options(
+                shape_invariants=[(tensor_to_copy, 
+                                   tensor_shape.TensorShape(None))]
+            )
+    lower_idx, upper_idx = paddings[i][0], paddings[i][1]
+    lower_begin = identity(start_coords)
+    upper_begin = identity(start_coords)
+    # Add the lower wrap.
+    while lower_idx > 0:
```

Comment:
I'm still not sure python while loops and if statements will work here in graph mode (i.e. `tf.function`).  It may work if `paddings` is constant, but if it's a variable tensor I'm not sure these would actually be evaluated correctly during tracing.

@faizan-m any experience here?

Reply:
I ran the function with `@tf.function`. You can see it working here: https://gist.github.com/madt2709/66fd514e08a159cb4ce04755bbde902e#file-wrap_pad-ipynb

Reply:
You can use `.get_concrete_function(...)` to get a function trace with variable inputs.
```
_wrap_pad_2_concrete = _wrap_pad_2.get_concrete_function(tf.TensorSpec(shape=[2, 2], dtype=tf.int32), tf.TensorSpec(shape=[2, 2], dtype=tf.int32))
_wrap_pad_2_concrete(t, paddings)
```
This actually seems to work though, so maybe it's fine.  I'll approve and import into Google to see if anything breaks.

Reply:
Hi!

Yeah tf.shape(tensor) will return a Tensor that has the dynamic shape and is always fully defined. tensor.shape will return the current static shape which can have undefined ranks/dimensions etc.

As Antonio mentioned the best way to be sure is to add a test where shape=None and try the result concrete function with specific inputs. If results are expected then you should be good.

Reply:
Thanks both! 

Just to be clear, should I add a test to explicitly check this or is this done by the internal CI?

Reply:
yeah that would be great!

Reply:
@faizan-m I added the extra test in. Lmk if it looks good!